azurecosmos/src/main/java/site/ycsb/db/AzureCosmosClient.java

```diff
@@ -187,218 +258,266 @@ private int getIntProperty(String propertyName, int defaultValue) {
-    ResourceResponse<Document> updatedResource = null;
-    ResourceResponse<Document> readResouce = null;
-    RequestOptions reqOptions = null;
-    Document document = null;
 
-    try {
-      reqOptions = getRequestOptions(key);
-      readResouce = AzureCosmosClient.client.readDocument(documentLink, reqOptions);
-      document = readResouce.getResource();
-    } catch (DocumentClientException e) {
-      if (!this.includeExceptionStackInLog) {
-        e = null;
-      }
-      LOGGER.error("Failed to read key {} in collection {} in database {} during update operation",
-          key, table, this.databaseName, e);
-      return Status.ERROR;
-    }
+    String readEtag = "";
 
-    // Update values
-    for (Entry<String, ByteIterator> entry : values.entrySet()) {
-      document.set(entry.getKey(), entry.getValue().toString());
-    }
+    for (int attempt = 0; attempt < NUM_UPDATE_ATTEMPTS; attempt++) {
 
-    AccessCondition accessCondition = new AccessCondition();
-    accessCondition.setCondition(document.getETag());
-    accessCondition.setType(AccessConditionType.IfMatch);
-    reqOptions.setAccessCondition(accessCondition);
+      try {
+        CosmosContainer container = AzureCosmosClient.containerCache.get(table);
+        if (container == null) {
+          container = AzureCosmosClient.database.getContainer(table);
+          AzureCosmosClient.containerCache.put(table, container);
+        }
 
-    try {
-      updatedResource = AzureCosmosClient.client.replaceDocument(documentLink, document, reqOptions);
-    } catch (DocumentClientException e) {
-      if (!this.includeExceptionStackInLog) {
-        e = null;
+        CosmosItemResponse<ObjectNode> response = container.readItem(key, new PartitionKey(key), ObjectNode.class);
+        readEtag = response.getETag();
+        ObjectNode node = response.getItem();
+
+        for (Entry<String, ByteIterator> pair : values.entrySet()) {
+          node.put(pair.getKey(), pair.getValue().toString());
+        }
+
+        CosmosItemRequestOptions requestOptions = new CosmosItemRequestOptions();
+        requestOptions.setIfMatchETag(readEtag);
+        PartitionKey pk = new PartitionKey(key);
+        container.replaceItem(node, key, pk, requestOptions);
+
+        return Status.OK;
+      } catch (CosmosException e) {
+        if (!AzureCosmosClient.includeExceptionStackInLog) {
+          e = null;
+        }
+        LOGGER.error("Failed to update key {} to collection {} in database {} on attempt {}", key, table,
+            AzureCosmosClient.databaseName, attempt, e);
+
+        // Azure Cosmos DB does not have patch support. Until then we need to read
+        // the document, update in place, and then write back.
+        // This could actually be made more efficient by using a stored procedure
+        // and doing the read/modify write on the server side. Perhaps
+        // that will be a future improvement.
```

Comment:
This comment does a better job of explaining thing when it precedes the code that does the read/modify/update cycle.